build_tools/build_detect_platform

```diff
@@ -827,6 +827,12 @@ fi
 
 rm -f test.o test_dl.o
 
+# Get the path for the folly installation dir
+if [ -z "$FOLLY_DIR" ]; then
+  FOLLY_DIR="./third-party/folly"
+fi
+FOLLY_PATH=`cd $FOLLY_DIR && $PYTHON build/fbcode_builder/getdeps.py show-inst-dir folly`
+
```

Comment:
Why not detect folly like any other add-on (such as ZSTD)?

Reply:
* Folly doesn't have proper versions with compatibility promises (see https://github.com/facebook/folly/issues/929#issuecomment-420820965)
* Folly uses a lot of conditional compilation and implementation-in-header, which likely affects linkability even between ABI-compatible compilers (because ODR, etc.)
* Folly tends to track newer C++ standards to take advantage of features

As a result of these things, pre-packaged folly binary+headers is not really a thing, and not expected to be in the foreseeable future.

Reply:
This is not pre-packaging.  This is "if I already have folly installed and am using it in one tool, why can I not use that same code I built?

CacheLib uses folly.  If there is a secondary cache that uses CacheLib, then RocksDB will end up with two versions of folly linked into it.  Why can't all of the tools use the same version of folly?

Reply:
OK, you've convinced me that we should think about this as an optional external dependency, potentially with a shortcut script to checkout & build a known-compatible version.

By the way our internal solution to this problem is "use buck."

Reply:
Makes sense. We'll have to detect not just folly, but its dependencies as well. If its not specified, then we can fall back to looking under FOLLY_DIR.